internal/cloud/azure/azure.go

```diff
@@ -409,6 +405,119 @@ func (c *Cloud) getVMInterfaces(ctx context.Context, vm armcompute.VirtualMachin
 	return networkInterfaces, nil
 }
 
+// getLoadBalancerPublicIP retrieves the first load balancer IP from cloud provider metadata.
+func (c *Cloud) getLoadBalancerPublicIP(ctx context.Context) (string, error) {
+	resourceGroup, err := c.imds.resourceGroup(ctx)
+	if err != nil {
+		return "", fmt.Errorf("retrieving resource group: %w", err)
+	}
+	uid, err := c.imds.uid(ctx)
+	if err != nil {
+		return "", fmt.Errorf("retrieving instance UID: %w", err)
+	}
+
+	lb, err := c.getLoadBalancer(ctx, resourceGroup, uid)
+	if err != nil {
+		return "", fmt.Errorf("retrieving load balancer: %w", err)
+	}
+	if lb == nil || lb.Properties == nil {
+		return "", errors.New("could not dereference load balancer IP configuration")
+	}
+
+	var pubIP string
+	for _, fipConf := range lb.Properties.FrontendIPConfigurations {
+		if fipConf == nil || fipConf.Properties == nil || fipConf.Properties.PublicIPAddress == nil || fipConf.Properties.PublicIPAddress.ID == nil {
+			continue
+		}
+		pubIP = path.Base(*fipConf.Properties.PublicIPAddress.ID)
+		break
+	}
+
+	resp, err := c.pubIPAPI.Get(ctx, resourceGroup, pubIP, nil)
+	if err != nil {
+		return "", fmt.Errorf("retrieving load balancer public IP address: %w", err)
+	}
+	if resp.Properties == nil || resp.Properties.IPAddress == nil {
+		return "", fmt.Errorf("could not resolve public IP address reference for load balancer")
+	}
+	return *resp.Properties.IPAddress, nil
+}
+
+// getLoadBalancerDNSWildcard retrieves dns wildcard of the load balancer.
+// For a public ip with DNS name "name-1234567890.location.cloudapp.azure.com",
+// this will return "*.location.cloudapp.azure.com".
+func (c *Cloud) getLoadBalancerDNSWildcard(ctx context.Context) (string, error) {
+	dnsName, err := c.getLoadBalancerDNSName(ctx)
+	if err != nil {
+		return "", fmt.Errorf("retrieving load balancer dns name: %w", err)
+	}
+	parts := strings.SplitN(dnsName, ".", 2)
+	if len(parts) != 2 {
+		return "", fmt.Errorf("invalid dns name %q", dnsName)
+	}
+	return fmt.Sprintf("*.%s", parts[1]), nil
+}
+
+// getLoadBalancerDNSName retrieves the dns name of the load balancer.
+// On Azure, the DNS name is the DNS name of the public IP address of the load balancer.
+func (c *Cloud) getLoadBalancerDNSName(ctx context.Context) (string, error) {
```

Comment:
This function is very similar to `getLoadBalancerPublicIP`.
Maybe combine some of the functionality into one function?

Reply:
I'm planning to remove `getLoadBalancerPublicIP` soon, as part if the second half of the canonical endpoint rfc. This is why most of the logic is duplicated.